swarm/swap_test.go

```diff
@@ -0,0 +1,423 @@
+//TestSwapNetworkAsymmetricFileUpload is a swap test too,
+//but this time the number and size of files are random
+func TestSwapNetworkAsymmetricFileUpload(t *testing.T) {
+	nodeCount := 16
+
+	sim := simulation.New(map[string]simulation.ServiceFunc{
+		"swarm": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+			config := api.NewConfig()
+			config.Port = strconv.Itoa(8500 + rand.Intn(9999))
+
+			dir, err := ioutil.TempDir("", "swap-network-test-node"+config.Port)
+			if err != nil {
+				return nil, nil, err
+			}
+			cleanup = func() {
+				err := os.RemoveAll(dir)
+				if err != nil {
+					log.Error("cleaning up swarm temp dir", "err", err)
+				}
+			}
+
+			config.Path = dir
+
+			privkey, err := crypto.GenerateKey()
+			if err != nil {
+				return nil, cleanup, err
+			}
+
+			config.Init(privkey)
+			//enable swap
+			config.SwapEnabled = true
+
+			swarm, err := NewSwarm(config, nil)
+			if err != nil {
+				return nil, cleanup, err
+			}
+			bucket.Store(bucketKeySwarm, swarm)
+			log.Info("new swarm", "bzzKey", config.BzzKey, "baseAddr", fmt.Sprintf("%x", swarm.bzz.BaseAddr()))
+			return swarm, cleanup, nil
+		},
+	})
+	defer sim.Close()
+
+	ctx := context.Background()
+	files := make([]file, 0)
+
+	var checkStatusM sync.Map
+	var nodeStatusM sync.Map
+	var totalFoundCount uint64
+
+	_, err := sim.AddNodesAndConnectChain(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	//NOTE: maxFileSize is 4 kB, this in order to provide faster tests
+	//it would be interesting to run these tests with bigger files
+	//(to see how drop limits are affected etc.)
+	const maxFileSize = 1024 * 4 //1024 bytes * 4 = 4kB
+	const minfileSize = 1024
+
+	//pseudo random algo to define if a node will upload or not
+	//if a bit is 0, do not upload
+	pseudoRandomNum := rand.Int63()
+	pseudoRandomBitMask := strconv.FormatInt(pseudoRandomNum, 2)
+
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		if _, err := sim.WaitTillHealthy(ctx, 2); err != nil {
+			return err
+		}
+
+		nodeIDs := sim.UpNodeIDs()
+		rand.Shuffle(len(nodeIDs), func(i, j int) {
+			nodeIDs[i], nodeIDs[j] = nodeIDs[j], nodeIDs[i]
+		})
+		for i, id := range nodeIDs {
+			//if the position in random num is 0, don't upload
+			if string(pseudoRandomBitMask[i]) != "0" {
+				size := rand.Intn(maxFileSize-minfileSize) + minfileSize
+				key, data, err := uploadRandomFileSize(sim.Service("swarm", id).(*Swarm), size)
+				if err != nil {
+					return err
+				}
+				log.Trace("file uploaded", "node", id, "key", key.String())
+				files = append(files, file{
+					addr:   key,
+					data:   data,
+					nodeID: id,
+				})
+			}
+		}
+
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		for {
+			if retrieve(sim, files, &checkStatusM, &nodeStatusM, &totalFoundCount) == 0 {
+				break
+			}
+		}
+
+		time.Sleep(5 * time.Second)
```

Comment:
?

Reply:
The sleep is there because we just finished the retrieval, but (accounted) messages are still floating around. 

If we start checking balances right away then the test fails.

5 seconds is a bit high but this was ensuring consistent results (consider travis). I will try with lower values.

Reply:
It works even with 1s Sleep (without Sleep still fails), but often I then get some amount of WARN messages like this:

`
WARN [11-30|07:09:02.918] ChunkStore.Get can not retrieve chunk    peer=979055cd185529b4ef2f618eb693885339bd84fa11f4126998c90eb4b9b14ac8 addr=4560305fc12373d47e345c32ac1b6823ca750653adab7eafb03620464b82f792 hopcount=1 err="context deadline exceeded"
`
or
`
WARN [11-30|07:13:48.785] SendOfferedHashes error                  err="leveldb: closed"
`

But the tests still pass. I have opted to keep `3s` for now but I am happy to change that.

Reply:
sleep usually means something is not right. 
In this case if some deliveries happen after if we retrieve all the chunks we need , it means we had to resort to another node. 
My suggestion is to make the test so that there is no delivery messages floating after we pass the retrieval. The best way to achieve this is to

* use a snapshot
* use a searchTimeout that is long
* use a requestTimeout that is as long as the searchTimeout.

Reply:
but `searchTimeout` is  private to the `swarm/network` package

Reply:
We have plenty of Sleeps in our tests BTW

Reply:
Latest commit has this fixed